File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int result = -1;
    
    public int diameterOfBinaryTree(TreeNode root) {
        // approach: get left height + right height at each node, bottom up
        
        helper(root);
        return result;
    }
    
    private int helper(TreeNode node) {
        if (node == null) {
            return -1;
        }
        
        int maxL = 1 + helper(node.left);
        int maxR = 1 + helper(node.right);
        result = Math.max(result, maxL + maxR);
        
        return Math.max(maxL, maxR);
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # approach: bottom-up DFS, get max of result and left height + right height
        
        result = -1
        
        def helper(node):
            nonlocal result
            
            if not node:
                return -1
            
            left = 1 + helper(node.left)
            right = 1 + helper(node.right)
            result = max(result, left + right)
            
            return max(left, right)
        
        helper(root)
        return result